### ISSUE TITLE
RFE: set the SCMP_FLTATR_CTL_TSYNC by default in the golang bindings

### ISSUE THREAD

**Issue by thorduri** [seccomp/libseccomp-golang#17]
Hi,

Due to the way the `go` runtime works (right off the bat on starts a handful of threads),
calling `SetTsync(true)` in `Load()` might reduce surprises for the package users.

Further, as a golang programmer, you have no control over thread creation / destruction or
scheduling (there are nuances here, but generally) so this doesn't map very well.

I'd go further, and suggest the API for setting/clearing this attribute should be removed
from the package and this just be done implicitly at Load time.

Prompted by https://github.com/opencontainers/runc/issues/1212

Comments ?

**Comment by pcmoore** [comment]
This seems perfectly reasonable to me, and isn't likely to cause a problem with any existing libseccomp-golang users due to the thread awareness limitations of Go.  @mheon what do you think?

**Comment by mheon** [comment]
General idea sounds fine here. You'd need to ignore potential error returns (library 2.1.x won't have the appropriate flag) when setting it in `Load` which is annoying --- it'd be nice if there were a way to identify that the flag was not set because of a too-early library version. Of course, we don't check if we're running on a kernel with `tsync` support anyways, so it's not that different from current behavior.

**Comment by pcmoore** [comment]
@thorduri @mheon okay, who wants to code this up into a patch/PR? ;)

**Comment by thorduri** [comment]
@pcmoore I can handle that.

However I think https://github.com/seccomp/libseccomp-golang/pull/15 would need to be merged first to address the issue pointed out by @mheon -- I've only verified libseccomp 2.3.1, however earlier versions might work (need to double check).

FWIW, docker et al, seem to "ship" 2.3.1, see e.g. [here](https://github.com/docker/docker/blob/master/Dockerfile#L112).

@mheon Do you think it would be worth it to gate this behaviour for kernels < 3.17 (which was released in Oct 2014 -- I don't) ?

**Comment by mheon** [comment]
@thorduri I wouldn't bother gating on the kernel. Perhaps a note in the documentation somewhere wouldn't be amiss, though. A note that we automatically enable tsync, but only if kernel and libseccomp are of sufficient version.

**Comment by pcmoore** [comment]
You can always just call seccomp_attr_set(SCMP_FLTATR_CTL_TSYNC) and if it succeeds then everything is okay, if not, well it's not supported :)

**Comment by mheon** [comment]
That's what we'll have to do. My concern is that we can't easily report that back to the calling application if we throw this in `Load`. Reporting an actual error is probably too heavy-handed (it did succeed, just without tsync), logging a warning is of questionable utility given the caller may have set up alternate logging infrastructure that the default Golang `log` package won't interact with.

**Comment by pcmoore** [comment]
Considering the fact that golang abstracts/hides threads, perhaps this isn't something we really need to worry about.

However, I might suggest attempting to set TSYNC immediately after calling seccomp_init() and not waiting until seccomp_load().  This way any golang callers that care deeply about this can query the TSYNC attribute via seccomp_attr_get() after creating a filter to see if it succeeded or not.

**Comment by thorduri** [comment]
The problem as I see it, is that without TSYNC the package doesn't do what it claims to on the tin.

Consider the following snippet (larger example [here](https://gist.github.com/thorduri/d66685860e7bafd93e573c53a36ef497)):
```go
        filter, err := seccomp.NewFilter(seccomp.ActErrno)
        if err != nil {
                log.Fatalf("seccomp.NewFilter: %v\n", err)
        }

        // note: getpid is not in the syscalls list.
        for _, syscall := range syscalls {
                allow(filter, syscall)
        }

        if err = filter.Load(); err != nil {
                fmt.Fprintf(os.Stderr, "error loading seccomp: %v\n", err)
                os.Exit(1)
        }

        http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
                fmt.Fprintf(w, "%d\n", os.Getpid())
        })

        log.Fatal(http.ListenAndServe(":8080", nil))
```

Failing to call (or not having support for) `SetTsync(true)` has a spectacular failure case, the call will succeed if the go routine that serves is not run on the same thread that ran the main code, e.g. if we run the larger example we get output similar to:

```
: ~ $; curl 127.0.0.1:8080/
0
: ~ $; curl 127.0.0.1:8080/
0
: ~ $; curl 127.0.0.1:8080/
3925
: ~ $; curl 127.0.0.1:8080/
3925
: ~ $; curl 127.0.0.1:8080/
0
```

This issue is present both in docker and libcontainerd, so I think the current API violates POLA.

Further, I think the minimum requirement for the binding should be set to v2.2.0, and the API to set/unset the bit should be removed. I'm proposing this in #18

**Comment by mheon** [comment]
I agree with your points - with the way Golang handles threads, not using `tsync` where available is just going to lead to misconfigurations which lead people to believe they're protected when there's a good chance they're not. The Docker usecase doesn't really feel this pain (Seccomp filters there are primarily to confine containers, not Docker itself - we configure/load the filters immediately before we `exec()` the program that's actually going to run in the container, so it doesn't really matter what our threads are doing, they'll disappear before anything important happens), but anyone intending on using an actual Go program with `libseccomp` filtering should be using `tsync`.

On raising the version requirement to v2.2: this is annoying for Docker, where they want to use Debian's v2.1.x, but they're building against the latest version now (and hoping that nothing blows up too horribly when they run with v2.1.1). Maybe this will force them to start recommending people install v2.2.x via PPA, which can only be a good thing for security (right now some of the default Docker Seccomp rules aren't enforced on Debian because of a bug in filter generation code in v2.1.1).

**Comment by thorduri** [comment]
@mheon Thanks for the explanation!

As to raising the requirement, I think its the right thing to do, and since docker does vendor in the bindings it will not cause a surprise downstream breakage (and v2.2.0 will soon be two years old...).

**Comment by kolyshkin** [comment]
@pcmoore @drakenclimber I apologize for mentioning you, but can you please take a look -- I beleive that as per previous comment, this issue can be closed.

I'm only looking at it because it's in v0.9.2 milestone, and I very much like v0.9.2 to be released, since the current situation where all libcseccomp-golang users have to rely on an unreleased version is not acceptable.

**Comment by drakenclimber** [comment]
> @pcmoore @drakenclimber I apologize for mentioning you, but can you please take a look -- I beleive that as per previous comment, this issue can be closed.

I agreed.  Looks like we've resolved this issue and it can be closed.  It only took five years :)

**Comment by kolyshkin** [comment]
For the sake of completeness -- this was actually fixed in April 2017 by commit fc0298087f328ac97a1fa781392de96b2ebb8aac.

@drakenclimber seems you forgot to close this.
